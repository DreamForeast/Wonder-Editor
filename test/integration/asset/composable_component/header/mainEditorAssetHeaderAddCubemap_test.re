open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("MainEditorAssetHeader->add cubemap", () => {
    let sandbox = getSandboxDefaultVal();

    let _prepareAndExecAndGetCubemapNode = () => {
      let assetTreeData =
        MainEditorAssetTreeTool.BuildAssetTree.Cubemap.buildOneCubemapAssetTree();
      let addedCubemapNodeId = MainEditorAssetIdTool.getNewAssetId();

      MainEditorAssetHeaderOperateNodeTool.addCubemap();

      (
        assetTreeData,
        StateEditorService.getState()
        |> OperateTreeAssetEditorService.unsafeFindNodeById(
             addedCubemapNodeId,
           ),
      );
    };

    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initState(~sandbox, ());

      MainEditorSceneTool.createDefaultSceneAndNotInit(sandbox);
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("if not select specific treeNode", () =>
      describe("should add cubemap into root treeNode", () => {
        test("test snapshot", () => {
          let assetTreeData =
            MainEditorAssetTreeTool.BuildAssetTree.Cubemap.buildOneCubemapAssetTree();

          MainEditorAssetHeaderOperateNodeTool.addCubemap();

          BuildComponentTool.buildAssetComponent()
          |> ReactTestTool.createSnapshotAndMatch;
        });
        test("the added cubemap's parent node should be root", () => {
          let (assetTreeData, node) = _prepareAndExecAndGetCubemapNode();

          MainEditorAssetTreeTool.findNodeParentId(
            node,
            StateEditorService.getState(),
          )
          |> OptionService.unsafeGet
          |> expect
          == MainEditorAssetTreeTool.BuildAssetTree.Cubemap.getRootNodeId(
               assetTreeData,
             );
        });
      })
    );

    describe("else", () =>
      describe("add cubemap into specific treeNode", () =>
        test("test snapshot", () => {
          let addedFolderNodeId1 = MainEditorAssetIdTool.getNewAssetId();
          MainEditorAssetHeaderOperateNodeTool.addFolder();

          MainEditorAssetTreeTool.Select.selectFolderNode(
            ~nodeId=addedFolderNodeId1,
            (),
          );
          MainEditorAssetHeaderOperateNodeTool.addCubemap();

          BuildComponentTool.buildAssetChildrenNode()
          |> ReactTestTool.createSnapshotAndMatch;
        })
      )
    );

    test("create new cubemap", () => {
      let assetTreeData =
        MainEditorAssetTreeTool.BuildAssetTree.Cubemap.buildOneCubemapAssetTree();
      let addedCubemapNodeId = MainEditorAssetIdTool.getNewAssetId();
      let newCubemap = CubemapTextureToolEngine.getNewCubemap();

      MainEditorAssetHeaderOperateNodeTool.addCubemap();

      let cubemapComponent =
        MainEditorAssetCubemapNodeTool.getCubemapTextureComponent(
          ~nodeId=addedCubemapNodeId,
          (),
        );

      cubemapComponent |> expect == newCubemap;
    });

    test("set face source imageDataIndex to None", () => {
      let assetTreeData =
        MainEditorAssetTreeTool.BuildAssetTree.Cubemap.buildOneCubemapAssetTree();
      let addedCubemapNodeId = MainEditorAssetIdTool.getNewAssetId();

      MainEditorAssetHeaderOperateNodeTool.addCubemap();

      let editorState = StateEditorService.getState();

      let {
        pxImageDataIndex,
        nxImageDataIndex,
        pyImageDataIndex,
        nyImageDataIndex,
        pzImageDataIndex,
        nzImageDataIndex,
      }: NodeAssetType.cubemapNodeData =
        OperateTreeAssetEditorService.unsafeFindNodeById(
          addedCubemapNodeId,
          editorState,
        )
        |> CubemapNodeAssetService.getNodeData;

      (
        pxImageDataIndex,
        nxImageDataIndex,
        pyImageDataIndex,
        nyImageDataIndex,
        pzImageDataIndex,
        nzImageDataIndex,
      )
      |> expect == (None, None, None, None, None, None);
    });

    describe("test name", () => {
      test("test default name", () => {
        let assetTreeData =
          MainEditorAssetTreeTool.BuildAssetTree.Cubemap.buildOneCubemapAssetTree();

        MainEditorAssetHeaderOperateNodeTool.addCubemap();

        BuildComponentTool.buildAssetComponent()
        |> ReactTestTool.createSnapshotAndMatch;
      });
      test(
        {|remove first cubemap which use default name;
             add three cubemap;

             the first new one's name should be removed-cubemap's name;
                     |},
        () => {
          let assetTreeData =
            MainEditorAssetTreeTool.BuildAssetTree.Cubemap.buildOneCubemapAssetTree();
          let nodeId =
            MainEditorAssetTreeTool.BuildAssetTree.Cubemap.getFirstCubemapNodeId(
              assetTreeData,
            );
          AssetTreeInspectorTool.Rename.renameAssetCubemapNode(
            ~nodeId,
            ~name=OperateCubemapLogicService.getDefaultName(),
            (),
          );

          MainEditorAssetHeaderOperateNodeTool.removeCubemapNode(
            ~cubemapNodeId=nodeId,
            (),
          );
          MainEditorAssetHeaderOperateNodeTool.addCubemap();
          MainEditorAssetHeaderOperateNodeTool.addCubemap();
          MainEditorAssetHeaderOperateNodeTool.addCubemap();
          MainEditorAssetTreeTool.Select.selectFolderNode(
            ~nodeId=
              MainEditorAssetTreeTool.BuildAssetTree.Cubemap.getRootNodeId(
                assetTreeData,
              ),
            (),
          );

          BuildComponentTool.buildAssetChildrenNode()
          |> ReactTestTool.createSnapshotAndMatch;
        },
      );
    });
  });