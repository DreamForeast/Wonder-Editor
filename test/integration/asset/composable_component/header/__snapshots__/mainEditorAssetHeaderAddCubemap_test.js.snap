// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainEditorAssetHeader->add cubemap else add cubemap into specific treeNode test snapshot 1`] = `
<article
  className="wonder-asset-assetChildren"
>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="./public/img/cubemap.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        New Cubemap
      </span>
    </div>
  </article>
</article>
`;

exports[`MainEditorAssetHeader->add cubemap if not select specific treeNode should add cubemap into root treeNode test snapshot 1`] = `
<article
  className="wonder-asset-component"
>
  <div
    className="asset-tree"
  >
    <article
      className="wonder-asset-header"
    >
      <div
        className="asset-header-item"
        onClick={[Function]}
        title="add asset"
      >
        <div
          className="item-canBeClick"
        >
          <img
            src="./public/img/add.png"
          />
        </div>
      </div>
      <div
        className="asset-header-item"
        onClick={[Function]}
        title="remove asset"
      >
        <div
          className="item-notBeClick"
        >
          <img
            src="./public/img/notRemove.png"
          />
        </div>
      </div>
      <div
        className="asset-header-item"
      >
        <div
          className="item-canBeClick"
          title="upload asset"
        >
          <img
            src="./public/img/load.png"
          />
          <input
            className="asset-fileLoad"
            multiple={false}
            onChange={[Function]}
            type="file"
          />
        </div>
      </div>
    </article>
    <article
      className="wonder-asset-assetTree"
    >
      <ul
        className="wonder-tree-node"
      >
        <li>
          <div
            className="item-triangle"
          />
          <img
            className="treeNode-icon"
            src="./public/img/package.png"
          />
          <div
            className="draggable-container select-not-active"
            draggable={true}
            onClick={[Function]}
            onDragEnd={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            style={Object {}}
          >
            Assets
          </div>
        </li>
      </ul>
    </article>
  </div>
  <article
    className="wonder-asset-assetChildren"
  >
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <div
        className="box-image"
      >
        <img
          onDragStart={[Function]}
          src="./public/img/cubemap.png"
        />
      </div>
      <div
        className="item-text "
      >
        <span>
          cubemap1
        </span>
      </div>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <div
        className="box-image"
      >
        <img
          onDragStart={[Function]}
          src="./public/img/cubemap.png"
        />
      </div>
      <div
        className="item-text "
      >
        <span>
          New Cubemap
        </span>
      </div>
    </article>
  </article>
</article>
`;

exports[`MainEditorAssetHeader->add cubemap test name remove first cubemap which use default name;
             add three cubemap;

             the first new one's name should be removed-cubemap's name;
                      1`] = `
<article
  className="wonder-asset-assetChildren"
>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="./public/img/cubemap.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        New Cubemap
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="./public/img/cubemap.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        New Cubemap 1
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="./public/img/cubemap.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        New Cubemap 2
      </span>
    </div>
  </article>
</article>
`;

exports[`MainEditorAssetHeader->add cubemap test name test default name 1`] = `
<article
  className="wonder-asset-component"
>
  <div
    className="asset-tree"
  >
    <article
      className="wonder-asset-header"
    >
      <div
        className="asset-header-item"
        onClick={[Function]}
        title="add asset"
      >
        <div
          className="item-canBeClick"
        >
          <img
            src="./public/img/add.png"
          />
        </div>
      </div>
      <div
        className="asset-header-item"
        onClick={[Function]}
        title="remove asset"
      >
        <div
          className="item-notBeClick"
        >
          <img
            src="./public/img/notRemove.png"
          />
        </div>
      </div>
      <div
        className="asset-header-item"
      >
        <div
          className="item-canBeClick"
          title="upload asset"
        >
          <img
            src="./public/img/load.png"
          />
          <input
            className="asset-fileLoad"
            multiple={false}
            onChange={[Function]}
            type="file"
          />
        </div>
      </div>
    </article>
    <article
      className="wonder-asset-assetTree"
    >
      <ul
        className="wonder-tree-node"
      >
        <li>
          <div
            className="item-triangle"
          />
          <img
            className="treeNode-icon"
            src="./public/img/package.png"
          />
          <div
            className="draggable-container select-not-active"
            draggable={true}
            onClick={[Function]}
            onDragEnd={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            style={Object {}}
          >
            Assets
          </div>
        </li>
      </ul>
    </article>
  </div>
  <article
    className="wonder-asset-assetChildren"
  >
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <div
        className="box-image"
      >
        <img
          onDragStart={[Function]}
          src="./public/img/cubemap.png"
        />
      </div>
      <div
        className="item-text "
      >
        <span>
          cubemap1
        </span>
      </div>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <div
        className="box-image"
      >
        <img
          onDragStart={[Function]}
          src="./public/img/cubemap.png"
        />
      </div>
      <div
        className="item-text "
      >
        <span>
          New Cubemap
        </span>
      </div>
    </article>
  </article>
</article>
`;
